src/nvim/message.c

```diff
@@ -1061,6 +1109,24 @@ static void msg_hist_add_multihl(HlMessage msg, bool temp)
   msg_hist_len += !temp;
   msg_hist_last = entry;
   msg_ext_history = true;
+
+  msg_ext_id = msg_id;
+  if (is_progress && msg_data != NULL && ui_has(kUIMessages)) {
+    kv_resize(msg_ext_progress, 3);
+    if (msg_data->title.size != 0) {
+      PUT_C(msg_ext_progress, "title", STRING_OBJ(msg_data->title));
+    }
+    if (msg_data->status.size != 0) {
+      PUT_C(msg_ext_progress, "status", STRING_OBJ(msg_data->status));
+    }
+    if (msg_data->percent >= 0) {
+      PUT_C(msg_ext_progress, "percent", INTEGER_OBJ(msg_data->percent));
+    }
+  }
+
+  if (is_progress) {
+    do_autocmd_progress(msg_id, entry, msg_data);
```

Comment:
not a blocker for this PR but hopefully can resolve it as a followup :

executing user code in random parts of the C core is risky and always leads to bugs.
is it possible to move this somewhere closer to where user code is *already* executed, later in this codepath perhaps? 

ideally by implementing https://github.com/neovim/neovim/issues/35366

Reply:
I'm not sure what you mean by later in codepath. It's already almost at the end. After this, we don't do much other then cleanup and function return.

A safe environment for executing autocommands would be nice. I believe it's out of scope for this pr.

Reply:
> After this, we don't do much other then cleanup and function return.

somewhere after the function return. i think the most sane approach is https://github.com/neovim/neovim/issues/35366 , but yes it's out of scope for now, though i don't think it will be hard.

Reply:
56cbe8c works but is not what justin was getting at. #35366 talks about queuing events and finding a safe state in the event loop to execute them (outside of the function calls that queue them).

Reply:
Yeah, that issue is a different thing. This pr is not trying to solve that issue. I was asking about the other mitigation justinmk was talking about : ]